### DISCUSSION TITLE
Cannot Get "Good" Shape for Tibetan Script

### DISCUSSION THREAD

**Discussion by yann-gael** [harfbuzz/harfbuzz#3548] [Q&A] [answered]
Hi!
(I hope this is the proper place to ask questions. If not, please let me know where to ask my questions and I'll be happy to forward it there.)
I'm new to font shaping and rendering. I followed the (useful) [simple shaping example](https://harfbuzz.github.io/a-simple-shaping-example.html) with some successes. I'm now trying to shape text in Tibetan and I cannot get "proper" glyphIDs. I'm getting as many 0-valued glyphID as code points in the input text. I have the same behaviour with _hb-shape_, built with HarfBuzz.
I think that I may have to add a shaping model but I just cannot figure out from the documentation what to create and how to use it. Could you please help me?
I post below my code if it can help.
Best,
Yann

My code:

#include <stdio.h>
#include <hb.h>

#define TEXT "\u0F68\u0F7C\u0F7E\u0F58\u0F4E\u0F72\u0F54\u0F51\u0FA8\u0F7A\u0F67\u0F71\u0F74\u0F83"
#define FONT "arial.ttf"

int main()
{
	hb_buffer_t *buf;
	buf = hb_buffer_create();
 	hb_buffer_add_utf8(buf, TEXT, -1, 0, -1);

	hb_buffer_set_direction(buf, HB_DIRECTION_LTR);
	hb_buffer_set_script(buf, HB_SCRIPT_TIBETAN);
	hb_buffer_set_language(buf, hb_language_from_string("bo", -1));

	hb_blob_t *blob = hb_blob_create_from_file(FONT);
	hb_face_t *face = hb_face_create(blob, 0);
	hb_font_t *font = hb_font_create(face);

	hb_shape(font, buf, NULL, 0);

	unsigned int glyph_count;
	hb_glyph_info_t *glyph_info    = hb_buffer_get_glyph_infos(buf, &glyph_count);
	hb_glyph_position_t *glyph_pos = hb_buffer_get_glyph_positions(buf, &glyph_count);

	hb_position_t cursor_x = 0;
	hb_position_t cursor_y = 0;
	for (unsigned int i = 0; i < glyph_count; i++) {
		hb_codepoint_t glyphid  = glyph_info[i].codepoint;
		hb_position_t x_advance = glyph_pos[i].x_advance;
		hb_position_t y_advance = glyph_pos[i].y_advance;

		printf("i = %02d, glyphID = %d\n", i, glyphid);

		cursor_x += x_advance;
		cursor_y += y_advance;
	}
	  
	hb_buffer_destroy(buf);
	hb_font_destroy(font);
	hb_face_destroy(face);
	hb_blob_destroy(blob);
	  
	return 0;
}

Current output is:

i = 00, glyphID = 0
i = 01, glyphID = 0
i = 02, glyphID = 0
i = 03, glyphID = 0
i = 04, glyphID = 0
i = 05, glyphID = 0
i = 06, glyphID = 0
i = 07, glyphID = 0
i = 08, glyphID = 0
i = 09, glyphID = 0
i = 10, glyphID = 0
i = 11, glyphID = 0
i = 12, glyphID = 0
i = 13, glyphID = 0

Expected output is: 6 non-0 glyphIDs.

**Comment by khaledhosny** [comment, marked as answer]
This means the font you are using, `arial.ttf`, does not support the specified characters.

**Reply by yann-gael** [reply]
Thank you for your reply. I thought as much but I can "see" the text in, for example, Microsoft Edge, which uses this font (as far as I can tell). So I tried with all the fonts used explicitly by Edge: Arial, Consola, and Times New Roman, with always the set results (set of 0-valued glyphID).

**Reply by behdad** [reply]
There's probably font fallback happening in Edge. Are you sure it's showing it using Arial?

**Reply by yann-gael** [reply]
Thank you for your help! Actually, no, I'm not sure: Edge doesn't say clearly which fonts its using.
However, I checked with Firefox, and it shows same behaviour as Edge, with same fonts explicitly defined.

**Reply by behdad** [reply]
All browsers support font fallback because that's required by CSS. You can see what font they are actually using if you select text then right-click and Inspect. It's in the Computed tab, at the very end under "Rendered Fonts", at least in Chrome. Should be the same in Edge and similar in Firefox.

**Reply by yann-gael** [reply]
I think that I must be doing something wrong, I just can't see what: I tried two simpler cases without success. 

Case 1:

#define TEXT "👩‍🦰"
#define FONT "times.ttf"
hb_buffer_set_script(buf, HB_SCRIPT_LATIN);
hb_buffer_set_language(buf, hb_language_from_string("en", -1));

produces:

i = 00, glyphID = 0
i = 01, glyphID = 3
i = 02, glyphID = 0

when I was expecting 1 non-0 glyphID.

Case 2: (code points for the same emoji above)

#define TEXT "\u1F469\u200D\u1F9B0"
#define FONT "times.ttf"
hb_buffer_set_script(buf, HB_SCRIPT_LATIN);
hb_buffer_set_language(buf, hb_language_from_string("en", -1));

produces:

i = 00, glyphID = 0
i = 01, glyphID = 28
i = 02, glyphID = 3
i = 03, glyphID = 2722
i = 04, glyphID = 19

when I was also expecting 1 non-0 glyphID (or at least the same output as in Case 1).

**Reply by yann-gael** [reply]
But your comments guided me into the right direction: it seems that Edge (and others?) use Segoe UI Emoji to render Emoji.

**Reply by behdad** [reply]
times.ttf DEFINITELY doesn't have emoji.  Your second example is also very suspect. I'm not sure why you are getting those glyph28 and 2722 19...  The glyphID=3 is expected, that corresponds to the U+200D in this case.

**Reply by yann-gael** [reply]
Thank you, you're of course right!

Using Segoe UI Emoji, I can get Case 1 to work as expected:

Case 3:

#define TEXT "👩‍🦰"
#define FONT "seguiemj.ttf"

finds 1 glyphID value 10849

However, Case 2 above doesn't produce the same glyphID:

Case 4:

#define TEXT "\u1F469\u200D\u1F9B0"
#define FONT "seguiemj.ttf"

i = 00, glyphID = 0
i = 01, glyphID = 28
i = 02, glyphID = 3
i = 03, glyphID = 0
i = 04, glyphID = 19

**Reply by behdad** [reply]
That means your compiler is not producing correct UTF-8 for "\u1F469".

**Reply by behdad** [reply]
Of-course. It's interpretting that as \u1F46, followed by digit "9".

**Reply by yann-gael** [reply]
Thanks! 
I'm using GCC 11.2.0 on CygWin, will check this out immediately.

**Reply by behdad** [reply]
Try \U0001F469 instead.

**Reply by yann-gael** [reply]
Wow, thanks!

Using

#define TEXT "\U0001F469\U0000200D\U0001F9B0"

did produce 1 glyphID with the same value as Case 3: 10849

It was of course my mistake: \u expects only 4 hex digits while \U up to eight.

**Comment by DavidLRowe** [comment]
You might want to try the Noto Sans Tibetan font, so that you are working with a font that supports the characters in the script.

**Reply by yann-gael** [reply]
Thanks!
I tried with himalaya.ttf provided with Windows 10 and NotoSerifTibetan-Regular.ttf that you suggested. I obtain 10 and 9 glyphIDs where I was expecting 6... but maybe I'm just misunderstanding how the Tibetan scripts works?

**Reply by yann-gael** [reply]
PS. I installed the extension FontFinder, which confirmed that himalaya.ttf is used by Edge to render Tibetan text.

**Reply by yann-gael** [reply]
PPS. I analysed himalaya.ttf using fontdrop.info and confirm that the 14 code points of ཨོཾམཎིཔདྨེཧཱུྃ are shaped into 11 glyphs with himalaya.ttf :-)
